map.c

```diff
@@ -920,6 +923,8 @@ static void *worker_pipeline(void *shared, int step, void *in)
 			s->trbuf = (mm_trbuf_t**)calloc(p->n_threads, sizeof(mm_trbuf_t*));
 			for (i = 0; i < p->n_threads; ++i)
 				s->trbuf[i] = mm_trbuf_init();
+			// TODO: initialize GPU infrastructures
+			// TODO: need a data structure to remember all the memptrs 
 #endif
```

Comment:
Confirm: s is shared globally (all worker threads will share the same s)
Confirm: s is multi-thread safe: s is inialized here and then it is read only to all the worker threads.

Reply:
`s` seems to be a global struct with many fields. One of those is `trbuf`, an array of `mm_trbuf_t` structure elements, one for each thread. So, `s` is not strictly readonly, but each thread has its own space to write to. That way, it is thread safe. Hope this makes sense.